pkg/core/native/notary.go

```diff
@@ -0,0 +1,345 @@
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("onPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(n.OnPersist), 0, smartcontract.AllowModifyStates)
+	n.AddMethod(md, desc, false)
+
+	desc = newDescriptor("postPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(postPersistBase), 0, smartcontract.AllowModifyStates)
+	n.AddMethod(md, desc, false)
+
+	return n
+}
+
+// Metadata implements Contract interface.
+func (n *Notary) Metadata() *interop.ContractMD {
+	return &n.ContractMD
+}
+
+// Initialize initializes Notary native contract and implements Contract interface.
+func (n *Notary) Initialize(ic *interop.Context) error {
+	return nil
+}
+
+// OnPersist implements Contract interface.
+func (n *Notary) OnPersist(ic *interop.Context) error {
+	var nFees int64
+	for _, tx := range ic.Block.Transactions {
+		if tx.HasAttribute(transaction.NotaryAssistedT) {
+			nkeys := tx.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+			nFees += int64(nkeys) + 1
+		}
+	}
+	if nFees == 0 {
+		return nil
+	}
+	notaries, err := n.GetNotaryNodes(ic.DAO)
+	if err != nil {
+		return fmt.Errorf("failed to get notary nodes: %w", err)
+	}
+	singleReword := calculateNotaryReword(nFees, len(notaries))
+	for _, notary := range notaries {
+		n.GAS.mint(ic, notary.GetScriptHash(), singleReword)
+	}
+	return nil
+}
+
+// onPayment records deposited amount as belonging to "from" address with a lock
+// till the specified chain's height.
+func (n *Notary) onPayment(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	amount := toBigInt(args[1])
+
+	till := defaultDepositLockedHeight
+	deposit := n.getDepositFor(ic.DAO, from)
+	if deposit == nil {
+		if amount.Cmp(big.NewInt(2*transaction.NotaryServiceFeeFactor)) < 0 {
+			return stackitem.Null{} // TODO: call ABORT instead
+		}
+		deposit = &state.Deposit{Amount: new(big.Int)}
+	} else {
+		till = deposit.Till
+	}
+	deposit.Amount.Add(deposit.Amount, amount)
+	deposit.Till = till
+	if err := n.putDepositFor(ic.DAO, deposit, from); err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", from.StringBE(), err))
+	}
+	return stackitem.Null{}
+}
+
+// lockDepositUntil updates the chain's height until which deposit is locked.
+func (n *Notary) lockDepositUntil(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	addr := toUint160(args[0])
+	ok, err := runtime.CheckHashedWitness(ic, addr)
+	if err != nil {
+		panic(fmt.Errorf("failed to check witness for %s: %w", addr.StringBE(), err))
+	}
+	if !ok {
+		return stackitem.NewBool(false)
+	}
+	till := uint32(toBigInt(args[1]).Int64())
+	if till < ic.Chain.BlockHeight() {
+		return stackitem.NewBool(false)
+	}
+	deposit := n.getDepositFor(ic.DAO, addr)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if till < deposit.Till { // TODO: do we really need this?
```

Comment:
There can be some transactions in-flight and this can allow withdrawing the deposit before they settle.